File: mobile/lib/utils/password_strength.dart
enum PasswordStrength { none, weak, medium, strong }

class PasswordStrengthEvaluator {
  PasswordStrengthEvaluator._();

  static final _letter = RegExp(r'[a-zA-Z]');
  static final _number = RegExp(r'\d');
  static final _special = RegExp(r'[!@#$%^&*(),.?":{}|<>_\-+=\[\]\\;/`~]');

  static PasswordStrength evaluate(String password) {
    if (password.isEmpty) return PasswordStrength.none;

    final hasLetter = _letter.hasMatch(password);
    final hasNumber = _number.hasMatch(password);
    final hasSpecial = _special.hasMatch(password);
    final categories = [hasLetter, hasNumber, hasSpecial].where((x) => x).length;

    if (categories >= 3) return PasswordStrength.strong;
    if (categories == 2) return PasswordStrength.medium;
    return PasswordStrength.weak;
  }

  static String label(PasswordStrength strength) {
    switch (strength) {
      case PasswordStrength.none:
        return '';
      case PasswordStrength.weak:
        return 'Weak - add letters, numbers, and symbols';
      case PasswordStrength.medium:
        return 'Fair - add a missing character type';
      case PasswordStrength.strong:
        return 'Strong password';
    }
  }

  static bool isStrong(String password) =>
      evaluate(password) == PasswordStrength.strong;
}

File: client/lib/password-strength.ts
export type PasswordStrength = 'none' | 'weak' | 'medium' | 'strong';

export const PASSWORD_STRONG_MESSAGE =
  'Use letters, numbers, and a special character';

const LETTER = /[a-zA-Z]/;
const NUMBER = /\d/;
const SPECIAL = /[!@#$%^&*(),.?":{}|<>_\-+=[\]\\;/`~]/;

export function evaluatePasswordStrength(password: string): PasswordStrength {
  if (!password) return 'none';

  const hasLetter = LETTER.test(password);
  const hasNumber = NUMBER.test(password);
  const hasSpecial = SPECIAL.test(password);
  const categories = [hasLetter, hasNumber, hasSpecial].filter(Boolean).length;

  if (categories >= 3) return 'strong';
  if (categories === 2) return 'medium';
  return 'weak';
}

export function passwordStrengthLabel(strength: PasswordStrength): string {
  switch (strength) {
    case 'none':
      return '';
    case 'weak':
      return 'Weak — add letters, numbers, and symbols';
    case 'medium':
      return 'Fair — add a missing character type';
    case 'strong':
      return 'Strong password';
  }
}

export function isStrongPassword(password: string): boolean {
  return evaluatePasswordStrength(password) === 'strong';
}

export function validatePasswordStrength(value: string): string | true {
  if (!value.trim()) return 'Password is required';
  return isStrongPassword(value) || PASSWORD_STRONG_MESSAGE;
}
